File: scripts/ci/check-critical-coverage.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";

const args = process.argv.slice(2);

function readArgValues(flag) {
  const values = [];
  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === flag && i + 1 < args.length) {
      values.push(args[i + 1]);
      i += 1;
    }
  }
  return values;
}

function readArgValue(flag, fallback) {
  const values = readArgValues(flag);
  return values.length > 0 ? values.at(-1) : fallback;
}

function parseNumber(value, fallback) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function normalizeFile(filePath) {
  const normalized = filePath.replace(/\\/g, "/").replace(/^\.\//, "");
  if (path.isAbsolute(normalized)) {
    return path.relative(process.cwd(), normalized).replace(/\\/g, "/");
  }
  return normalized;
}

function toPercent(covered, total) {
  if (!total) {
    return 100;
  }
  return (covered / total) * 100;
}

function globToRegExp(glob) {
  const escaped = glob
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replace(/\*\*/g, "__DOUBLE_STAR__")
    .replace(/\*/g, "[^/]*")
    .replace(/__DOUBLE_STAR__/g, ".*")
    .replace(/\?/g, ".");
  return new RegExp(`^${escaped}$`);
}

function matchesPattern(filePath, pattern) {
  if (!pattern.includes("*") && !pattern.includes("?")) {
    return filePath === pattern || filePath.endsWith(`/${pattern}`);
  }
  return globToRegExp(pattern).test(filePath);
}

function parseLcov(lcovPath) {
  const raw = fs.readFileSync(lcovPath, "utf8");
  const lines = raw.split(/\r?\n/);
  const fileMap = new Map();
  let currentFile = null;

  for (const line of lines) {
    if (line.startsWith("SF:")) {
      currentFile = normalizeFile(line.slice(3).trim());
      if (!fileMap.has(currentFile)) {
        fileMap.set(currentFile, {
          lineCovered: 0,
          lineTotal: 0,
          branchCovered: 0,
          branchTotal: 0,
        });
      }
      continue;
    }
    if (!currentFile) {
      continue;
    }
    if (line.startsWith("DA:")) {
      const payload = line.slice(3).trim().split(",");
      if (payload.length < 2) {
        continue;
      }
      const hits = Number(payload[1]);
      const item = fileMap.get(currentFile);
      item.lineTotal += 1;
      if (Number.isFinite(hits) && hits > 0) {
        item.lineCovered += 1;
      }
      continue;
    }
    if (line.startsWith("BRDA:")) {
      const payload = line.slice(5).trim().split(",");
      if (payload.length < 4) {
        continue;
      }
      const taken = payload[3];
      const item = fileMap.get(currentFile);
      item.branchTotal += 1;
      if (taken !== "-" && Number.isFinite(Number(taken)) && Number(taken) > 0) {
        item.branchCovered += 1;
      }
      continue;
    }
    if (line === "end_of_record") {
      currentFile = null;
    }
  }

  let lineCovered = 0;
  let lineTotal = 0;
  let branchCovered = 0;
  let branchTotal = 0;
  for (const value of fileMap.values()) {
    lineCovered += value.lineCovered;
    lineTotal += value.lineTotal;
    branchCovered += value.branchCovered;
    branchTotal += value.branchTotal;
  }

  return {
    type: "lcov",
    path: lcovPath,
    lineCovered,
    lineTotal,
    branchCovered,
    branchTotal,
    files: fileMap,
  };
}

function readCoverageAttribute(xml, name) {
  const match = xml.match(new RegExp(`${name}="([0-9]+(?:\\.[0-9]+)?)"`));
  if (!match) {
    return null;
  }
  const parsed = Number(match[1]);
  return Number.isFinite(parsed) ? parsed : null;
}

function parseCobertura(xmlPath) {
  const raw = fs.readFileSync(xmlPath, "utf8");
  const linesValid = readCoverageAttribute(raw, "lines-valid");
  const linesCovered = readCoverageAttribute(raw, "lines-covered");
  const branchesValid = readCoverageAttribute(raw, "branches-valid");
  const branchesCovered = readCoverageAttribute(raw, "branches-covered");

  const classMatches = Array.from(
    raw.matchAll(
      /<class[^>]*filename="([^"]+)"[^>]*line-rate="([0-9.]+)"(?:[^>]*branch-rate="([0-9.]+)")?[^>]*>/g,
    ),
  );
  const files = new Map();

  for (const match of classMatches) {
    const fileName = normalizeFile(match[1]);
    const lineRate = Number(match[2]);
    const branchRate = Number(match[3]);
    const lineTotal = 100;
    const branchTotal = 100;
    files.set(fileName, {
      lineCovered: Math.round((Number.isFinite(lineRate) ? lineRate : 0) * lineTotal),
      lineTotal,
      branchCovered: Math.round((Number.isFinite(branchRate) ? branchRate : 1) * branchTotal),
      branchTotal,
    });
  }

  return {
    type: "cobertura",
    path: xmlPath,
    lineCovered: linesCovered ?? 0,
    lineTotal: linesValid ?? 0,
    branchCovered: branchesCovered ?? 0,
    branchTotal: branchesValid ?? 0,
    files,
  };
}

function fail(message) {
  console.error(`[critical-coverage-gate] ${message}`);
  process.exit(1);
}

const lcovPaths = readArgValues("--lcov");
const coberturaPaths = readArgValues("--cobertura");
const criticalPatterns = readArgValues("--critical");
const globalThresholdLegacy = readArgValue("--global-threshold", undefined);
const criticalThresholdLegacy = readArgValue("--critical-threshold", undefined);

if (lcovPaths.length === 0 && coberturaPaths.length === 0) {
  fail("missing coverage input. Provide at least one --lcov or --cobertura file.");
}

if (criticalPatterns.length === 0) {
  fail("missing --critical patterns. Provide at least one critical module pattern.");
}

const globalLineThreshold = parseNumber(
  readArgValue("--global-line-threshold", globalThresholdLegacy ?? "85"),
  85,
);
const criticalLineThreshold = parseNumber(
  readArgValue("--critical-line-threshold", criticalThresholdLegacy ?? "95"),
  95,
);
const globalBranchThreshold = parseNumber(readArgValue("--global-branch-threshold", "70"), 70);
const criticalBranchThreshold = parseNumber(readArgValue("--critical-branch-threshold", "90"), 90);

const reports = [];
for (const filePath of lcovPaths) {
  if (!fs.existsSync(filePath)) {
    fail(`lcov file not found: ${filePath}`);
  }
  reports.push(parseLcov(filePath));
}
for (const filePath of coberturaPaths) {
  if (!fs.existsSync(filePath)) {
    fail(`cobertura file not found: ${filePath}`);
  }
  reports.push(parseCobertura(filePath));
}

let totalLineCovered = 0;
let totalLineCount = 0;
let totalBranchCovered = 0;
let totalBranchCount = 0;
for (const report of reports) {
  totalLineCovered += report.lineCovered;
  totalLineCount += report.lineTotal;
  totalBranchCovered += report.branchCovered;
  totalBranchCount += report.branchTotal;
}

const globalLinePercent = toPercent(totalLineCovered, totalLineCount);
const globalBranchPercent = toPercent(totalBranchCovered, totalBranchCount);

const allFiles = new Map();
for (const report of reports) {
  for (const [filePath, metrics] of report.files.entries()) {
    if (!allFiles.has(filePath)) {
      allFiles.set(filePath, {
        lineCovered: 0,
        lineTotal: 0,
        branchCovered: 0,
        branchTotal: 0,
      });
    }
    const aggregate = allFiles.get(filePath);
    aggregate.lineCovered += metrics.lineCovered;
    aggregate.lineTotal += metrics.lineTotal;
    aggregate.branchCovered += metrics.branchCovered;
    aggregate.branchTotal += metrics.branchTotal;
  }
}

const failures = [];
const criticalResults = [];
for (const pattern of criticalPatterns) {
  const matches = Array.from(allFiles.entries()).filter(([filePath]) =>
    matchesPattern(filePath, pattern),
  );
  if (matches.length === 0) {
    failures.push(`critical pattern unmatched: ${pattern}`);
    continue;
  }
  for (const [filePath, metrics] of matches) {
    const linePercent = toPercent(metrics.lineCovered, metrics.lineTotal);
    const branchPercent = toPercent(metrics.branchCovered, metrics.branchTotal);
    criticalResults.push({ pattern, filePath, linePercent, branchPercent });
    if (linePercent + 1e-9 < criticalLineThreshold) {
      failures.push(
        `critical line coverage below threshold: ${filePath} (${linePercent.toFixed(2)}% < ${criticalLineThreshold.toFixed(2)}%)`,
      );
    }
    if (branchPercent + 1e-9 < criticalBranchThreshold) {
      failures.push(
        `critical branch coverage below threshold: ${filePath} (${branchPercent.toFixed(2)}% < ${criticalBranchThreshold.toFixed(2)}%)`,
      );
    }
  }
}
for (const result of criticalResults.sort((a, b) => a.filePath.localeCompare(b.filePath))) {
  console.log(
    `[critical-coverage-gate] critical ${result.filePath}: line ${result.linePercent.toFixed(2)}% / branch ${result.branchPercent.toFixed(2)}% [pattern: ${result.pattern}]`,
  );
}

if (globalLinePercent + 1e-9 < globalLineThreshold) {
  failures.push(
    `global line coverage below threshold: ${globalLinePercent.toFixed(2)}% < ${globalLineThreshold.toFixed(2)}%`,
  );
}
if (globalBranchPercent + 1e-9 < globalBranchThreshold) {
  failures.push(
    `global branch coverage below threshold: ${globalBranchPercent.toFixed(2)}% < ${globalBranchThreshold.toFixed(2)}%`,
  );
}

console.log(
  `[critical-coverage-gate] global line ${globalLinePercent.toFixed(2)}% (threshold ${globalLineThreshold.toFixed(2)}%)`,
);
console.log(
  `[critical-coverage-gate] global branch ${globalBranchPercent.toFixed(2)}% (threshold ${globalBranchThreshold.toFixed(2)}%)`,
);

if (failures.length > 0) {
  console.error("[critical-coverage-gate] FAILED");
  for (const item of failures) {
    console.error(`- ${item}`);
  }
  process.exit(1);
}

console.log("[critical-coverage-gate] PASSED");

File: scripts/ci/check-critical-coverage.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import fnmatch
import pathlib
import sys
import xml.etree.ElementTree as ET


def _norm(path: str) -> str:
    p = path.replace("\\\\", "/").lstrip("./")
    return p


def _match(path: str, pattern: str) -> bool:
    if "*" not in pattern and "?" not in pattern:
        return path == pattern or path.endswith("/" + pattern)
    return fnmatch.fnmatch(path, pattern)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Cobertura critical coverage gate")
    parser.add_argument("--coverage-file", required=True)
    # Backward compatibility: legacy line-only flags.
    parser.add_argument("--global-threshold", type=float, default=None)
    parser.add_argument("--critical-threshold", type=float, default=None)
    parser.add_argument("--global-line-threshold", type=float, default=None)
    parser.add_argument("--global-branch-threshold", type=float, default=70.0)
    parser.add_argument("--critical-line-threshold", type=float, default=None)
    parser.add_argument("--critical-branch-threshold", type=float, default=90.0)
    parser.add_argument("--critical", action="append", default=[])
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if not args.critical:
        print("[critical-coverage-gate] missing --critical patterns", file=sys.stderr)
        return 1

    global_line_threshold = (
        args.global_line_threshold
        if args.global_line_threshold is not None
        else (args.global_threshold if args.global_threshold is not None else 85.0)
    )
    critical_line_threshold = (
        args.critical_line_threshold
        if args.critical_line_threshold is not None
        else (args.critical_threshold if args.critical_threshold is not None else 95.0)
    )

    coverage_path = pathlib.Path(args.coverage_file)
    if not coverage_path.exists():
        print(
            f"[critical-coverage-gate] cobertura file not found: {coverage_path}", file=sys.stderr
        )
        return 1

    root = ET.parse(coverage_path).getroot()
    lines_valid = float(root.attrib.get("lines-valid", "0") or 0)
    lines_covered = float(root.attrib.get("lines-covered", "0") or 0)
    branches_valid = float(root.attrib.get("branches-valid", "0") or 0)
    branches_covered = float(root.attrib.get("branches-covered", "0") or 0)
    global_line_percent = 100.0 if lines_valid == 0 else (lines_covered / lines_valid) * 100.0
    global_branch_percent = (
        100.0 if branches_valid == 0 else (branches_covered / branches_valid) * 100.0
    )

    classes: dict[str, tuple[float, float]] = {}
    for class_node in root.findall(".//class"):
        filename = class_node.attrib.get("filename", "")
        line_rate = float(class_node.attrib.get("line-rate", "0") or 0)
        branch_rate = float(class_node.attrib.get("branch-rate", "1") or 1)
        if filename:
            classes[_norm(filename)] = (line_rate * 100.0, branch_rate * 100.0)

    failures: list[str] = []
    if global_line_percent < global_line_threshold:
        failures.append(
            f"global line coverage below threshold: {global_line_percent:.2f}% < {global_line_threshold:.2f}%"
        )
    if global_branch_percent < args.global_branch_threshold:
        failures.append(
            f"global branch coverage below threshold: {global_branch_percent:.2f}% < {args.global_branch_threshold:.2f}%"
        )

    matched_rows: list[tuple[str, str, float, float]] = []
    for pattern in args.critical:
        matches = [(path, metrics) for path, metrics in classes.items() if _match(path, pattern)]
        if not matches:
            failures.append(f"critical pattern unmatched: {pattern}")
            continue
        for path, (line_pct, branch_pct) in matches:
            matched_rows.append((pattern, path, line_pct, branch_pct))
            if line_pct < critical_line_threshold:
                failures.append(
                    f"critical line coverage below threshold: {path} ({line_pct:.2f}% < {critical_line_threshold:.2f}%)"
                )
            if branch_pct < args.critical_branch_threshold:
                failures.append(
                    f"critical branch coverage below threshold: {path} ({branch_pct:.2f}% < {args.critical_branch_threshold:.2f}%)"
                )

    print("[critical-coverage-gate] Coverage summary")
    print(f"- global line: {global_line_percent:.2f}% (threshold {global_line_threshold:.2f}%)")
    print(
        f"- global branch: {global_branch_percent:.2f}% (threshold {args.global_branch_threshold:.2f}%)"
    )
    print(f"- total lines: {int(lines_covered)}/{int(lines_valid)}")
    print(f"- total branches: {int(branches_covered)}/{int(branches_valid)}")
    print("- critical modules:")
    for pattern, path, line_pct, branch_pct in sorted(matched_rows, key=lambda row: row[1]):
        print(f"  - {path}: line {line_pct:.2f}% / branch {branch_pct:.2f}% [pattern: {pattern}]")

    if failures:
        print("[critical-coverage-gate] FAILED", file=sys.stderr)
        for item in failures:
            print(f"- {item}", file=sys.stderr)
        return 1

    print("[critical-coverage-gate] PASSED")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
